### ISSUE TITLE
Automatic selection of fastest OpenCL platform doesn't work if there is more than one platform and a specific device is selected

### ISSUE THREAD

**Issue by thtrummer** [openmm/openmm#2712]
[This line](https://github.com/openmm/openmm/blob/6b39ec7ec0285e1584531efa8d56fc571e0d0c9f/platforms/opencl/src/OpenCLContext.cpp#L95) fails when there is more than one OpenCL platform, `OPENCL_TEST_PLATFORM_INDEX` is set to -1 (default) and `OPENCL_TEST_DEVICE_INDEX` is overriden by a test.

This happens for example [here](https://github.com/openmm/openmm/blob/6b39ec7ec0285e1584531efa8d56fc571e0d0c9f/platforms/opencl/tests/TestOpenCLCheckpoints.cpp#L100) where `deviceIndex == "0"`.

**Comment by peastman** [comment]
That's intentional.  If you specify a device index, that indicates there's a particular device you want to use.  But if you don't specify which of the multiple platforms you're referring to, we don't actually know which device you mean.  This almost certainly indicates user error, so we throw an exception.

**Comment by thtrummer** [comment]
This makes sense. However, it still means that the test is broken, since it must specify the platform of the device instead of using the value of `OPENCL_TEST_PLATFORM_INDEX` (which here is `-1`)?

**Comment by peastman** [comment]
Run the test as

```
./TestOpenCLCheckpoints single 0
```

You can pass up to three arguments to OpenCL tests.  In order, they're the precision, the platform index, and the device index.

**Comment by thtrummer** [comment]
I've figured out as much. Let me rephrase the problem: The [documentation](http://docs.openmm.org/7.1.0/userguide/library.html#step-6-test-your-build) claims that you can call `make test` to run all tests. This currently doesn't work because all tests that use two devices will fail with an exception (when there is more than one platform).

There is missing something (e.g. [here](https://github.com/openmm/openmm/blob/6b39ec7ec0285e1584531efa8d56fc571e0d0c9f/platforms/opencl/tests/TestOpenCLCheckpoints.cpp#L100)) along the lines of:

```cpp
string platformIndex = platform.getPropertyValue(context, OpenCLPlatform::OpenCLPlatformIndex());
props[OpenCLPlatform::OpenCLPlatformIndex()] = platformIndex;
```

**Comment by peastman** [comment]
I see what you mean.  Yes, that would be a reasonable change.  (This situation doesn't come up often, because it's rare to have two different platforms installed.)

**Comment by thtrummer** [comment]
Fixed by #2717.
